WordPress/openverse-catalog
# Add docs and template ProviderDataIngester

## Fixes


Fixes #757 by @stacimc

## Description


* Adds a `create_provider_ingester` script that can be used to generate a templated provider api script and test file. The generated files include a lot of documentation and TODOs aimed at helping a new contributor flesh out an ingestion class. Credit to an [earlier template script in this repository](https://github.com/WordPress/openverse-catalog/tree/1c9fd4cf46f7c90f8e9db6eaacb31d6ad031fcf5/openverse_catalog/templates), from which I lifted quite a bit of code!
* Adds the following docs (would love better title suggestions 😅):
  * `adding_a_new_provider.md`: Details a very brief overview of what it even means to add a provider to Openverse, explains how to use the script, and also a brief explanation of how to add a `ProviderWorkflow` to generate the actual DAG
  * `provider_data_ingester_faq.md`: Describes "advanced options" for implementing some common non-standard use cases in a ProviderDataIngester, framed as an FAQ
  * `data_models.md`: Extremely temporary documentation of our columns, meant to be replaced by #783.
* Expands some of the documentation in the actual `ProviderDataIngester` class

The intention is to avoid too much duplication of documentation from the _code_. If deemed necessary, we can look into auto-generating some from the doc strings, but my strategy was:
* Script makes it easy to generate a stubbed out new ingester class
* Comments and TODOs within the generated class lead the developer through implementing a "straightforward" provider
* "Advanced" options are documented separately with examples

There's definitely a balancing act of where to document things in the code vs the docfile vs the template. My goal is mostly for this to be a good starting point!

## Testing Instructions


* Read the docs , make sure links work, etc :)
* Test the new script
  * `just test`
    * Verify that the test files (`foobar_industries.py` & `test_foobar_industries.py`) were deleted after the tests finished
  * Run it with some sample data, ie: `just add-provider "Foo Museum" "https://foo-museum.org/api/v1/" image audio`
    * Verify the `foo_museum.py` provider script and `test_foo_museum.py` files were generated
    * Inspect the files to make sure they make sense and that the name/endpoint/media types were templated in properly
    * Try again with different media types, including a single media type
    

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
